### ISSUE TITLE
go-spacemesh/api-phase9 grpcserver.MeshService.readLayer error

### ISSUE THREAD

**Issue by AndrewAR2** [spacemeshos/go-spacemesh#2084]
## Description
Error into grpcserver.MeshService.readLayer:

```
2020-07-28T08:17:58.519+0300<-->INFO<-->00000.defaultLogger<--->reported new layer: {0xc005563f00 0}
2020-07-28T08:17:58.519+0300<-->INFO<-->00000.defaultLogger<--->reported status update
2020-07-28T08:17:58.519+0300<-->INFO<-->00000.defaultLogger<--->not reporting new layer as no one is listening: {0xc005a90040 0}
2020-07-28T08:17:58.528+0300<-->ERROR<->00000.defaultLogger<--->error getting state root for layer 10
2020-07-28T08:17:58.529+0300<-->INFO<-->00000.defaultLogger<--->not reporting error as buffer is full: {error getting state root for layer 10 goroutine 534 [running]:
runtime/debug.Stack(0x0, 0xc005d7e2d0, 0xc0044074a8)
<------>/home/merlin/src/goroot/src/runtime/debug/stack.go:24 +0x9d
github.com/spacemeshos/go-spacemesh/cmd/node.(*SpacemeshApp).setupLogging.func1(0x2, 0xbfc00b219f835d96, 0x4ab8dc277, 0x19fb420, 0xffe85b, 0x13, 0xc0059b26f0, 0x25, 0x0, 0x0, ...)
<------>/home/merlin/go/src/github.com/spacemeshos/go-spacemesh/cmd/node/node.go:291 +0x42
go.uber.org/zap/zapcore.(*hooked).Write(0xc000465530, 0x2, 0xbfc00b219f835d96, 0x4ab8dc277, 0x19fb420, 0xffe85b, 0x13, 0xc0059b26f0, 0x25, 0x0, ...)
<------>/home/merlin/go/pkg/mod/go.uber.org/zap@v1.10.0/zapcore/hook.go:65 +0x9b
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc003d3a420, 0x0, 0x0, 0x0)
<------>/home/merlin/go/pkg/mod/go.uber.org/zap@v1.10.0/zapcore/entry.go:215 +0x117
go.uber.org/zap.(*SugaredLogger).log(0xc00040e318, 0xc0058f7202, 0x1015978, 0x25, 0xc004407978, 0x1, 0x1, 0x0, 0x0, 0x0)
<------>/home/merlin/go/pkg/mod/go.uber.org/zap@v1.10.0/sugar.go:234 +0x100
go.uber.org/zap.(*SugaredLogger).Errorf(...)
<------>/home/merlin/go/pkg/mod/go.uber.org/zap@v1.10.0/sugar.go:148
github.com/spacemeshos/go-spacemesh/log.Log.Error(...)
<------>/home/merlin/go/src/github.com/spacemeshos/go-spacemesh/log/zap.go:36
github.com/spacemeshos/go-spacemesh/log.Error(...)
<------>/home/merlin/go/src/github.com/spacemeshos/go-spacemesh/log/log.go:150
github.com/spacemeshos/go-spacemesh/api/grpcserver.MeshService.readLayer(0x11bf180, 0xc000103c80, 0xc0003fc330, 0x7f0640271f80, 0xc00060e540, 0x6, 0x2, 0xc8, 0x1e, 0x64, ...)
<------>/home/merlin/go/src/github.com/spacemeshos/go-spacemesh/api/grpcserver/mesh_service.go:326 +0xc7c
github.com/spacemeshos/go-spacemesh/api/grpcserver.MeshService.LayerStream(0x11bf180, 0xc000103c80, 0xc0003fc330, 0x7f0640271f80, 0xc00060e540, 0x6, 0x2, 0xc8, 0x1e, 0x64, ...)
<------>/home/merlin/go/src/github.com/spacemeshos/go-spacemesh/api/grpcserver/mesh_service.go:485 +0x1be
github.com/spacemeshos/api/release/go/spacemesh/v1._MeshService_LayerStream_Handler(0xfc5380, 0xc000450550, 0x11b2c00, 0xc003d9e0c0, 0x1a286c0, 0xc000164900)
<------>/home/merlin/go/pkg/mod/github.com/spacemeshos/api/release/go@v0.0.0-20200720193236-49b94a501b3e/spacemesh/v1/mesh.pb.go:638 +0x109
google.golang.org/grpc.(*Server).processStreamingRPC(0xc002ec4000, 0x11b87e0, 0xc0043cea80, 0xc000164900, 0xc00016a990, 0x18edb80, 0x0, 0x0, 0x0)
<------>/home/merlin/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:1329 +0xcbd
google.golang.org/grpc.(*Server).handleStream(0xc002ec4000, 0x11b87e0, 0xc0043cea80, 0xc000164900, 0x0)
<------>/home/merlin/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:1409 +0xc66
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc003bb2270, 0xc002ec4000, 0x11b87e0, 0xc0043cea80, 0xc000164900)
<------>/home/merlin/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:746 +0xa1
created by google.golang.org/grpc.(*Server).serveStreams.func1
<------>/home/merlin/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:744 +0xa1
 2}
```

**Comment by lrettig** [comment]
Thanks. Could you share the command you're using to run the node and the config file? Is it the devnet config file I posted on slack yesterday?

**Comment by lrettig** [comment]
Actually a full log file would be very helpful for debugging in case you can share it. Thanks!

**Comment by AndrewAR2** [comment]
> Thanks. Could you share the command you're using to run the node and the config file? 

./go-spacemesh --grpc node,mesh,globalstate,transaction,smesher --config ./config-dev-net.json -d ./sm_data

> Is it the devnet config file I posted on slack yesterday?

Yes!

**Comment by AndrewAR2** [comment]
> Actually a full log file would be very helpful for debugging in case you can share it. Thanks!

I'll try it, but here's a fairly short log from start to the error:

```
2020-07-28T08:17:38.482+0300	INFO	00000.defaultLogger	App version: v0.0.0-unreleased. Git: api-phase9 - 876305c359408d49ad69123f8067407dc26a0f6c . Go Version: go1.14.2. OS: linux-amd64 
2020-07-28T08:17:38.482+0300	INFO	00000.defaultLogger	initializing event reporter
2020-07-28T08:17:38.482+0300	INFO	00000.defaultLogger	Welcome to Spacemesh. Spacemesh full node is starting...
2020-07-28T08:17:38.869+0300	INFO	00000.defaultLogger	ntp check from server time-b-wwv.nist.gov, drift: -77.10598ms, start_check: 2020-07-28 05:17:38.711667764 +0000 UTC, latency: 126.783µs, rsp: 2020-07-28 05:17:38.788710353 +0000 UTC
2020-07-28T08:17:38.909+0300	INFO	00000.defaultLogger	ntp check from server time1.google.com, drift: -19.521935ms, start_check: 2020-07-28 05:17:38.871647876 +0000 UTC, latency: 74.243µs, rsp: 2020-07-28 05:17:38.89113269 +0000 UTC
2020-07-28T08:17:38.945+0300	INFO	00000.defaultLogger	ntp check from server time.asia.apple.com, drift: -18.551008ms, start_check: 2020-07-28 05:17:38.911679921 +0000 UTC, latency: 57.807µs, rsp: 2020-07-28 05:17:38.930202026 +0000 UTC
2020-07-28T08:17:38.985+0300	INFO	00000.defaultLogger	ntp check from server time4.google.com, drift: -19.840192ms, start_check: 2020-07-28 05:17:38.947515386 +0000 UTC, latency: 73.577µs, rsp: 2020-07-28 05:17:38.96731879 +0000 UTC
2020-07-28T08:17:39.143+0300	INFO	00000.defaultLogger	ntp check from server time-a-wwv.nist.gov, drift: -76.630376ms, start_check: 2020-07-28 05:17:38.987265831 +0000 UTC, latency: 77.196µs, rsp: 2020-07-28 05:17:39.063857609 +0000 UTC
2020-07-28T08:17:39.143+0300	INFO	00000.defaultLogger	System clock synchronized with ntp. drift: -38.664167ms
2020-07-28T08:17:39.143+0300	INFO	00000.defaultLogger	Starting Spacemesh	{"data-dir": "sm_data/2", "post-dir": "/home/merlin/post/data"}
2020-07-28T08:17:39.143+0300	INFO	00000.defaultLogger	Loaded identity from file ('/home/merlin/post/data/51fcc622307cc7f4d0cc8421a34825e38704d39f5407e9a406211c88c923cddb/key.bin')
2020-07-28T08:17:39.158+0300	INFO	00000.defaultLogger	Initializing P2P services
2020-07-28T08:17:39.159+0300	INFO	clock	starting global clock now=2020-07-28 08:17:39.158887396 +0300 EEST m=+0.688248125 genesis=2020-07-23 15:40:28.369 +0000 +0000
2020-07-28T08:17:39.159+0300	INFO	clock	global clock going to sleep before next layer	{"diff": "2m49.209680849s", "next_layer": 1974}
2020-07-28T08:17:39.164+0300	INFO	00000.defaultLogger	Saved p2p node information (sm_data/2/p2p/nodes/HtmDpqAGrJWZCo3cmivzgnsKLbQ9Tjh4b1XLgaDizgT9/id.json),  Identity - HtmDpqAGrJWZCo3cmivzgnsKLbQ9Tjh4b1XLgaDizgT9
2020-07-28T08:17:39.165+0300	INFO	00000.defaultLogger	Loaded 17 addresses from file 'sm_data/2/p2p/peers.json'
2020-07-28T08:17:39.165+0300	INFO	00000.defaultLogger	not reporting status update as no one is listening
2020-07-28T08:17:39.376+0300	INFO	00000.defaultLogger	started processor with state root 56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421
2020-07-28T08:17:39.395+0300	INFO	00000.defaultLogger	not reporting status update as no one is listening
2020-07-28T08:17:39.395+0300	INFO	00000.defaultLogger	not reporting status update as no one is listening
2020-07-28T08:17:39.395+0300	INFO	00000.defaultLogger	subscribed to channel
2020-07-28T08:17:39.395+0300	INFO	00000.defaultLogger	subscribed to channel
2020-07-28T08:17:39.396+0300	INFO	00000.defaultLogger	not reporting status update as no one is listening
2020-07-28T08:17:39.396+0300	INFO	00000.defaultLogger	Manual post init
2020-07-28T08:17:39.396+0300	INFO	ticker	started notifying
2020-07-28T08:17:39.396+0300	INFO	00000.defaultLogger	challenge not loaded: leveldb: not found
2020-07-28T08:17:44.783+0300	INFO	00000.defaultLogger	not reporting status update as no one is listening
2020-07-28T08:17:44.783+0300	INFO	00000.defaultLogger	starting new grpc server with 5 registered service(s)
2020-07-28T08:17:44.783+0300	INFO	00000.defaultLogger	subscribed to channel
2020-07-28T08:17:44.783+0300	INFO	00000.defaultLogger	App started.
2020-07-28T08:17:44.783+0300	INFO	00000.defaultLogger	starting new grpc server on :9092
2020-07-28T08:17:46.663+0300	INFO	51fcc.p2p          	discovery_bootstrap	{"P2PID": "HtmDpqAGrJWZCo3cmivzgnsKLbQ9Tjh4b1XLgaDizgT9", "event": true, "success": true, "size": 30, "time_elapsed": "1.880533225s"}
2020-07-28T08:17:46.807+0300	INFO	00000.defaultLogger	not reporting status update as no one is listening
2020-07-28T08:17:46.962+0300	INFO	00000.defaultLogger	GRPC MeshService.GenesisTime
2020-07-28T08:17:46.966+0300	INFO	00000.defaultLogger	GRPC MeshService.EpochNumLayers
2020-07-28T08:17:46.967+0300	INFO	00000.defaultLogger	GRPC MeshService.MaxTransactionsPerSecond
2020-07-28T08:17:46.968+0300	INFO	00000.defaultLogger	GRPC GlobalStateService.GlobalStateStream
2020-07-28T08:17:46.968+0300	INFO	00000.defaultLogger	GRPC MeshService.LayerStream
2020-07-28T08:17:46.969+0300	INFO	00000.defaultLogger	GRPC NodeService.StatusStream
2020-07-28T08:17:46.969+0300	INFO	00000.defaultLogger	GRPC NodeService.SyncStart
2020-07-28T08:17:49.396+0300	INFO	00000.defaultLogger	reported status update
2020-07-28T08:17:51.557+0300	INFO	00000.defaultLogger	reported new layer: {0xc00369a360 0}
2020-07-28T08:17:51.557+0300	INFO	00000.defaultLogger	reported status update
2020-07-28T08:17:51.557+0300	INFO	00000.defaultLogger	not reporting new layer as no one is listening: {0xc00369a4e0 0}
2020-07-28T08:17:51.557+0300	ERROR	00000.defaultLogger	error getting state root for layer 8
2020-07-28T08:17:51.557+0300	INFO	00000.defaultLogger	not reporting error as buffer is full: {error getting state root for layer 8 goroutine 310 [running]:
runtime/debug.Stack(0x0, 0xc003199020, 0xc003d834a8)
	/home/merlin/src/goroot/src/runtime/debug/stack.go:24 +0x9d
github.com/spacemeshos/go-spacemesh/cmd/node.(*SpacemeshApp).setupLogging.func1(0x2, 0xbfc00b1fe13a7c26, 0x30c095b1f, 0x19fb420, 0xffe85b, 0x13, 0xc0036c07e0, 0x24, 0x0, 0x0, ...)
	/home/merlin/go/src/github.com/spacemeshos/go-spacemesh/cmd/node/node.go:291 +0x42
go.uber.org/zap/zapcore.(*hooked).Write(0xc000465530, 0x2, 0xbfc00b1fe13a7c26, 0x30c095b1f, 0x19fb420, 0xffe85b, 0x13, 0xc0036c07e0, 0x24, 0x0, ...)
	/home/merlin/go/pkg/mod/go.uber.org/zap@v1.10.0/zapcore/hook.go:65 +0x9b
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc006d2e000, 0x0, 0x0, 0x0)
	/home/merlin/go/pkg/mod/go.uber.org/zap@v1.10.0/zapcore/entry.go:215 +0x117
go.uber.org/zap.(*SugaredLogger).log(0xc00040e318, 0xc000113702, 0x1015978, 0x25, 0xc003d83978, 0x1, 0x1, 0x0, 0x0, 0x0)
	/home/merlin/go/pkg/mod/go.uber.org/zap@v1.10.0/sugar.go:234 +0x100
go.uber.org/zap.(*SugaredLogger).Errorf(...)
	/home/merlin/go/pkg/mod/go.uber.org/zap@v1.10.0/sugar.go:148
github.com/spacemeshos/go-spacemesh/log.Log.Error(...)
	/home/merlin/go/src/github.com/spacemeshos/go-spacemesh/log/zap.go:36
github.com/spacemeshos/go-spacemesh/log.Error(...)
	/home/merlin/go/src/github.com/spacemeshos/go-spacemesh/log/log.go:150
github.com/spacemeshos/go-spacemesh/api/grpcserver.MeshService.readLayer(0x11bf180, 0xc000103c80, 0xc0003fc330, 0x7f0640271f80, 0xc00060e540, 0x6, 0x2, 0xc8, 0x1e, 0x64, ...)
	/home/merlin/go/src/github.com/spacemeshos/go-spacemesh/api/grpcserver/mesh_service.go:326 +0xc7c
github.com/spacemeshos/go-spacemesh/api/grpcserver.MeshService.LayerStream(0x11bf180, 0xc000103c80, 0xc0003fc330, 0x7f0640271f80, 0xc00060e540, 0x6, 0x2, 0xc8, 0x1e, 0x64, ...)
	/home/merlin/go/src/github.com/spacemeshos/go-spacemesh/api/grpcserver/mesh_service.go:485 +0x1be
github.com/spacemeshos/api/release/go/spacemesh/v1._MeshService_LayerStream_Handler(0xfc5380, 0xc000450550, 0x11b2c00, 0xc004372000, 0x1a286c0, 0xc000412600)
	/home/merlin/go/pkg/mod/github.com/spacemeshos/api/release/go@v0.0.0-20200720193236-49b94a501b3e/spacemesh/v1/mesh.pb.go:638 +0x109
google.golang.org/grpc.(*Server).processStreamingRPC(0xc002ec4000, 0x11b87e0, 0xc000672480, 0xc000412600, 0xc00016a990, 0x18edb80, 0x0, 0x0, 0x0)
	/home/merlin/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:1329 +0xcbd
google.golang.org/grpc.(*Server).handleStream(0xc002ec4000, 0x11b87e0, 0xc000672480, 0xc000412600, 0x0)
	/home/merlin/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:1409 +0xc66
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc00067b870, 0xc002ec4000, 0x11b87e0, 0xc000672480, 0xc000412600)
	/home/merlin/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:746 +0xa1
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/home/merlin/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:744 +0xa1
 2}
2020-07-28T08:17:51.562+0300	INFO	00000.defaultLogger	StatusStream closing stream, client disconnected
2020-07-28T08:17:51.562+0300	INFO	00000.defaultLogger	GlobalStateStream closing stream, client disconnected

```

**Comment by lrettig** [comment]
I looked into this today. I was able to recreate it—it seems to happen every time the node receives new blocks from peers, while someone is listening on LayerQuery. Good catch, thank you for reporting this.

I suspect it's just happening because the layer data is received/streamed before the layer state root is recorded in the database. I need to dig a little bit deeper and determine whether this is expected behavior, i.e., whether it should be an error in the first place.
